P1684-mdarray/P1684.md

```diff
@@ -644,6 +651,504 @@ The larger or equal is intentional, because the mapping may already be not exhau
+    they just need to make sure that the `vector`'s lifetime
+    contains the lifetimes of all the `mdspan` that view its elements.
+
+#### "Reshape" reuses storage and elements
+
+Reusing storage and elements
+is equivalent to the "reshape" facility
+found in other programming languages and libraries.
+This facility reinterprets an existing array
+as an array with a different (multidimensional) "shape."
+Examples include Fortran's `RESHAPE` intrinsic function.
+The `mdarray` class cannot be "reshaped in place,"
+because the extents and layout are part of the type.
+Even if all the extents are dynamic, the *rank* is static.
+
+If `mdarray` had a facility for extracting the container,
+then one way to implement an "out-of-place reshape"
+would be to extract the container from the source `mdarray`
+and move it into a new destination `mdarray`.
+However, depending on the container type,
+this may copy the elements instead of viewing the original elements.
+An example of such a container type is `std::array`.
+
+Thus, the only way to implement a reshape operation
+that *always* views the original elements
+would be to use `mdspan`.  That is, the reshape operation
+could view an existing `mdarray` as an `mdspan`
+with different extents and/or layout.
+
+### Effects of extracting the container from an mdarray
+
+The discussion above explains why we do not think
+`mdarray` needs an operation for extracting its container.
+If we were nevertheless to add such an operation,
+this section explains its postconditions:
+that is, the state of the "extracted-from" `mdarray`.
+
+#### An "extracted-from" mdarray is like a moved-from object
+
+The hypothetical `extract_container()` member function
+would have the effect of returning `move(ctr_)`.
+If the caller uses the result in a `container_type`
+move constructor or move assignment operator,
+then this would leave the `mdarray`'s container in a moved-from state.
+Given that `mdarray` can accept any contiguous container
+and imposes no requirements on the container's moved-from state,
+the resulting moved-from container would be in a
+"valid but unspecified" state ([lib.types.movedfrom]).
+This would make it invalid to access any elements
+of the "extracted-from" `mdarray`.
+
+If the `mdarray` has any dynamic extents,
+these could hypothetically be set to zero, thus ensuring that
+the `mdarray`'s multidimensional index space would be empty.
+However, if the `mdarray` has all static extents,
+there is no way to zero those out,
+as they are part of the `mdarray`'s type.
+The `mdarray` would be unusable.
+Its destructor would still do the right thing,
+but users could no longer rely on
+any other properties of the `mdarray`.
+It would effectively be in a "valid but unspecified state,"
+just like its container.
+
+This means that generic code should treat
+an extracted-from `mdarray` like any moved-from object:
+the only safe operation is to let it be destroyed.
```

Comment:
From a language point of view, I agree that objects need to be destructible without any preconditions.

From a library point of view, one can also assume that a type in a moved-from state can be assigned-to without any preconditions.  Is there a reason we don't say this?

And to be clear, neither of those is necessarily  a "valid but unspecified" state.  It can only be that if such a state maintains the class invariants, either directly or by adding preconditions on other operations.  This is sometimes referred to as a "partially formed state".

(Okay, it may not matter here since we are recommending not to have this functionality, but it does matter for the moved-from state of `mdarray`.)

Reply:
@nliber Thanks for pointing this out!  

> From a library point of view, one can also assume that a type in a moved-from state can be assigned-to without any preconditions. Is there a reason we don't say this?

Is this part of the contiguous container requirements, or is this a general property of all C++ objects?  What in the Standard says that a moved-from object needs to be assignable?